proposals/NNNN-package-access-modifier.md

```diff
@@ -0,0 +1,176 @@
+
+```
+[Game]
+
+import Engine
+
+public func play() {
+    MainEngine().run() // Can access `run` as it is a package symbol in the same package
+}
+```
+
+However, if a client outside of the package tries to access the helper API, it will not be allowed.
+
+```
+[App]
+
+import Game
+import Engine
+
+let engine = MainEngine()
+engine.run() // Error: cannot find `run` in scope
+
+```
+
+During type check a package name that's stored in an imported module binary will be looked up and compared with the current module's package name to allow or disallow access to a package symbol from that module. More details are explained below.
+
+
+### Lookup by package name
+
+A new flag `-package-name` will be introduced to enable grouping modules per package. In Swift Package Manager, the input to the flag will be a package identity and will be passed down automatically. Each package identity is unique and can contain non-alphanumeric characters such as a hyphen and a dot (currently any URL valid characters are allowed); such characters will be transposed into a c99 identifier. Other build systems such as Xcode and Basel will need to pass a new command-line argument `-package-name` to the build command; the input may be Reverse-DNS package names so it can prevent potential clashes with other project or package names. 
+
+```
+[Engine] swiftc -module-name Engine -package-name gamePkg ...
+[Game] swiftc -module-name Game -package-name gamePkg ...
+[App] swiftc App -package-name appPkg ...
+```
+
+When building the Engine module, the package name 'gamePkg' and package symbols will be stored in the Engine.swiftmodule. When building Game, the input to its package name 'gamePkg' will be compared with Engine's package name; since they match, access to package symbols is allowed. When building App, the name comparison shows 'appPkg' is different from its dependency module's so access to package symbols will be deined, which is what we want.
+
+If `-package-name` is not passed, but `package` access modifier is used, then an error will be thrown; Building files that do not have `package` symbols should continue to work without needing to pass in `-package-name`. 
+
+
+### Exportability
+
+The package symbols will be stored in .swiftmodule but not in .swiftinterface. The package name will be stored in both .swiftmodule and .swiftinterface, but only the swiftmodule will be loaded when building a same-package module; this will prevent loading a fall-back module (swiftinterface) which does not contain package symbols.  
+
+The exportability rule at the declaration site will be similar to the the existing behavior; for example, a public class is not allowed to inherit a package class, and a public func is not allowed to have a package type in its signature. 
+
+The `@inlinable` will be applicable to `package`; we plan to introduce`@usableFromPackageInline` to allow references from `@inlinable package` and not from `@inlinable public`.
```

Comment:
```suggestion
`package` functions can be made `@inlinable`.  Just like with `@inlinable public`, not all symbols are usable within the function: they must be `open`, `public`, `package`, or `@usableFromInline`.  Note that `@usableFromInline` allows the use of a symbol from `@inlinable` functions whether they're `package` or `public`; this proposal does not contain a way to export a symbol only for use by `@inlinable package` functions, although we plan to add one in a future proposal.
```

Although, I've got to say that that's a pretty narrow follow-up proposal.  Have you considered just including this now?

Reply:
I can include that in this proposal.